askama_derive/src/generator.rs

```diff
@@ -638,18 +642,30 @@ impl<'a> Generator<'a> {
                             use super::input::EscapeMode::*;
                             let mut expr_buf = Buffer::new(0);
                             let wrapped = self.visit_expr(&mut expr_buf, s);
-
-                            buf_format.write("{}");
-                            buf_expr.write("&");
-                            buf_expr.write(&match (wrapped, &self.input.escaping) {
-                                (Wrapped, &Html) | (Wrapped, &None) | (Unwrapped, &None) => {
-                                    expr_buf.buf
-                                }
-                                (Unwrapped, &Html) => {
-                                    format!("::askama::MarkupDisplay::from(&{})", expr_buf.buf)
+                            if cache_id.contains_key(&expr_buf) {
+                                let id = cache_id.get(&expr_buf).unwrap();
+                                buf_format.write(&format!("{{a{}}}", id));
+                            } else {
+                                let id = self.named;
+                                buf_expr.write(&format!("a{}=", id));
+                                buf_expr.write("&");
+                                match (wrapped, &self.input.escaping) {
+                                    (Wrapped, &Html) | (Wrapped, &None) | (Unwrapped, &None) => {
+                                        buf_expr.write(&expr_buf.buf);
+                                    }
+                                    (Unwrapped, &Html) => {
+                                        buf_expr.write(&format!(
+                                            "::askama::MarkupDisplay::from(&{})",
+                                            expr_buf.buf
+                                        ));
+                                    }
                                 }
-                            });
-                            buf_expr.writeln(",");
+
+                                buf_expr.writeln(",");
+                                buf_format.write(&format!("{{a{}}}", id));
+                                cache_id.insert(expr_buf.clone(), id);
```

Comment:
Move the `insert()` and `named += 1` up to right after the line `let id = self.named` so the logic is better localized?

Actually, we can probably move ownership of `expr_buf` into the cache map? In that case, it needs to be down at the bottom, but we should get rid of the `clone()` (and the corresponding `derive()`).

Reply:
and ok